src/lsm/eytzinger.zig

```diff
@@ -183,7 +186,24 @@ pub fn eytzinger(comptime keys_count: u32, comptime values_max: u32) type {
 
             var i: u32 = 0;
             while (i < keys.len) {
-                // TODO use @prefetch when available: https://github.com/ziglang/zig/issues/3600
+                if (comptime prefetch) {
+                    // Prefetching is only worthy if we can fit many keys into the cache line.
+                    assert(constants.cache_line_size > @sizeOf(Key));
+
+                    // Prefetching next levels from memory.
+                    // https://arxiv.org/abs/1509.05053
+                    const cache_line_keys =
+                        comptime @divExact(constants.cache_line_size, @sizeOf(Key));
+
+                    // We need to use pointer arithmetic to avoid bounds checks.
+                    // Locality = 0 means no temporal locality. That is, the data can be immediately
+                    // dropped from the cache after it is accessed.
+                    @prefetch(keys.ptr + (cache_line_keys * i) + cache_line_keys, .{
```

Comment:
> Can you document why we diverge from the paper's `⌊3B/2⌋ − 1` offset? Is this just faster experimentally, or is this derived somehow?

For reference:

This [Rust project](https://github.com/bazhenov/eytzinger-layout) implements it as `ptr + 2 * i`, but differs from us because our `i` starts at zero.
https://github.com/bazhenov/eytzinger-layout/blob/77be9ae4b0d4eaa753e4bb180150ca77fc4d5239/src/lib.rs#L36-L39

Algorithmica.org implements it as `ptr + i * block_size`, but again, it differs from us by starting with i=1.
https://algorithmica.org/en/eytzinger#prefetching

And the paper implements it as `ptr + (i * cache_line)` but inexplicably (at least to me) adds an offset `(3 * cache_line / 2) - 1`.

> Also to be clear, are we trying to prefetch for the current iteration, or for the _next_ iteration, when the `while` loop wraps around?

The idea is to prefetch the next iteration.

> I would assume it should be the latter, but since there's no `2 * ...` in the prefetch statement that doesn't appear to be the case.

I think I need to build a stronger mental model on how `2 * i` would relate to `cache_line_keys` before being able to answer that!

Reply:
Yeah, I don't understand the paper's `⌊3B/2⌋ − 1` offset at all. (Admittedly it has been a long time since I read the full paper, and I only skimmed it when reviewing this PR. But I couldn't find an derivation or explanation for the offset.)

> I think I need to build a stronger mental model on how 2 * i would relate to cache_line_keys before being able to answer that!

My intuition here is:

Immediately before the loop wraps around, we update `i`:

```zig
i = if (compare_keys(key, keys[i]) == .gt) 2 * i + 2 else 2 * i + 1;
```

So during the next iteration, `i` is either `2 * i + 2` or `2 * i + 1`.
Ideally, we would prefetch a cache line that contains both of those.
(And if there is room left over in the cache line, we would like it to be on the "right" side, since we might jump forward again afterwards.)

---

Also, unrelated, could you add a note to the `i = ...` line that this branchy version is actually faster than the branchless version? (Along with the "why", if known.)

Reply:
Pat Morin did a talk explaining all the bitwise tricks in more detail, and I think he may have explained the prefetch calculation in there too. I'm trying to find it...

In the meantime, here's the talk by Paul Khuong, co-author. I haven't watched this, but sharing so long in case it has it: https://www.youtube.com/watch?v=n5IwFC_LRp0

Reply:
The last section of the paper (6.2) goes into when branchy can be faster.

Reply:
For posterity: https://www.youtube.com/watch?v=-1qRcDnbihA

Reply:
> I think I need to build a stronger mental model on how 2 * i would relate to cache_line_keys before being able to answer that!

@sentientwaffle please check the explanation on the code comments.
I'm 0.9 sure that I got it right, the 0.1 left is the reasoning behind `⌊3B/2⌋ − 1`, anyways, empirically using this offset was slower both for micro-benchmarking and for the real workload.